File: solutions/0034-find-first-and-last-position-of-element-in-sorted-array/solution.java
class Solution {
    static int binarySearching(int[] arr, int target){
        //for array sorted in ascending order only.
        int high = arr.length-1;
        int low = 0;
        while(high>=low){
            int mid = low+(high-low)/2;
            if(arr[mid]==target) return mid;
            else if(arr[mid]>target) high = mid-1;
            else low = mid+1;
        }
        return -1;
    }
    public static int[] searchRange(int[] nums, int target) {
        int[] arr = new int[2];
        int currIndex = binarySearching(nums, target);
        if(currIndex==-1){
            arr[0] = -1;
            arr[1] = -1;
        }
        else {
            int first = currIndex, last = currIndex;
            while (first > 0) {
                if (nums[first - 1] == nums[currIndex]){
                    first--;
                    arr[0] = first;
                }
                else {
                    arr[0] = first;
                    break;
                }
            }
            if(last== nums.length-1){
                arr[1] = last;
            }
            else {
                while (last < nums.length - 1) {
                    if (nums[last + 1] == nums[currIndex]) {
                        last++;
                        arr[1] = last;
                    } else {
                        arr[1] = last;
                        break;
                    }
                }
            }
        }
        return arr;
    }
}

File: solutions/0034-find-first-and-last-position-of-element-in-sorted-array/solution.cpp
class Solution {
public:
    vector<int> search(vector<int>& nums, int targetIndex){
        int ptr1 = targetIndex;
        int ptr2 = targetIndex;
        while(ptr1>=0&&ptr2<nums.size()){
            if(nums[ptr1]!=nums[targetIndex]&&nums[ptr2]!=nums[targetIndex]) break;
            if(nums[ptr1]==nums[targetIndex]){
                ptr1--;
            }
            if(nums[ptr2]==nums[targetIndex]){
                ptr2++;
            }
        }
        while(ptr2<nums.size()){
            if(nums[ptr2]!=nums[targetIndex]) break;
            if(nums[ptr2]==nums[targetIndex]){
                ptr2++;
            }
        }
        while(ptr1>=0){
            if(nums[ptr1]!=nums[targetIndex]) break;
            if(nums[ptr1]==nums[targetIndex]){
                ptr1--;
            }
        }
        return {ptr1+1, ptr2-1};
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = 0;
        int h = nums.size()-1;
        vector<int> res(2);
        while(l<=h){
            int m = l+(h-l)/2;
            if(nums[m]==target){
                cout<<m<<" ";
                return search(nums, m);
            }
            else if(nums[m]<target) l = m+1;
            else h = m-1;
        }
        return {-1, -1};
    }
};
